Use (hand-rolled) sparse matrix multiplication

astralblue opened the pull request:
This replaces use of `gonum.org/v1/gonum/mat.(*Dense)` and `gonum.org/v1/gonum/mat.(*VecDense)` with a hand-rolled sparse matrix/vector implementation.  Memory usage is now O(NNZ + R), where NNZ is the number of non-zero entries and R is the number of rows.

astralblue committed:
Add TimeLogger to log a series of elapsed times

astralblue committed:
Instrument (*StrictServerImpl).Compute

astralblue committed:
Use (hand-rolled) sparse matrix multiplication

This reduces the memory usage to be on the order of non-zero entries of
trust matrix/vectors, enabling Lens use case.

astralblue committed:
Pre-allocate sparse entries during Transpose()

Reallocations in append() calls are pretty expensive,
and they happen a lot.
They also lead to memory spikes due to deferred garbage collections.

astralblue committed:
Use 32 goroutines to parallelize MulVec()

astralblue committed:
Remove unused debug variable

astralblue committed:
Log Compute() progress/stats

astralblue committed:
Log input dimensions/NNZ

(NNZ = number of non-zero entries in a sparse matrix/vector.)

astralblue committed:
Fix CLI: Make -p/--pre-trust of basic compute work

astralblue committed:
Actually set the context logger in server

astralblue committed:
Allow LT/PT to have different dimensions

That is, scale the smaller one up to match the bigger one.

This is easily doable now that we have a sparse implementation.

astralblue committed:
Move Err(ZeroSum|DimensionMismatch) to pkg/sparse

astralblue committed:
Make Dim() return an error; don't panic

astralblue committed:
[playground] Do not panic on dimension mismatch

astralblue committed:
Replace Grow...() with Set...Dim()

astralblue committed:
Replace Rows()/Columns() with Dims()

This is more in line with existing conventions (ex: gonum).

astralblue committed:
Do not panic in pkg/sparse

astralblue committed:
Nit: columns -> cols (in line with gonum)

astralblue committed:
Use the correct dimension for MulVec() prechecks

v is the out parameter; v1 should be used for input checks.

astralblue committed:
MulVec: entryCh/entries -> entries/sortedEntries

astralblue committed:
Make Transpose() and MulVec() interruptible

These two tend to take the longest time.

astralblue committed:
Fix tests: Leave zero-length slice at nil

astralblue committed:
Log Compute progress using logarithmic distance

log10dPace is the movement of d,
expressed as log10(d/d0) where d0 is the previous d
(or on first iteration, d0=1.0 - maximum possible Norm2).

log10dRemaining is the remaining logarithmic distance,
expressed as log10(d/e); Compute() finishes when this drops below 0.

This is easier to quickly grasp

astralblue committed:
Move logger helpers to pkg/util

astralblue committed:
Fix AddVec/SubVec

astralblue committed:
Remove lap time logging from server

astralblue committed:
Fix reversed log10d logging

astralblue committed:
Publish sortEntries -> SortEntries

web123pro commented on pkg/sparse/entry.go:
```diff
@@ -0,0 +1,71 @@
+func (a entrySort) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+
+func SortEntries(entries []Entry) []Entry {
+	entries = append(entries[:0:0], entries...)
```
Could we have a TODO to change this method to do an in-place sort ? If the calling code needs a new copy to be sorted, then it create a copy before calling this function.

astralblue replied:
Fixed! <3

astralblue committed:
Garbage collect after each iteration & request

web123pro commented on pkg/basic/server.go:
```diff
@@ -152,7 +182,7 @@ func loadInlineTrustVector(inline *InlineTrustVector) (TrustVector, error) {
+		entries = append(entries, sparse.Entry{Index: entry.I, Value: entry.V})
 	}
-	return lt, nil
+	return sparse.NewVector(inline.Size, entries), nil
```
At this point in the code, we would have 3 copies of the TrustVector - 1 copy in the InlineTrustVector, 2nd copy in entries slice, and  3rd copy in sparse.NewVector.

astralblue replied:
Fixed.  Now we have only two copies at most at any given point in time, and only one copy after this function returns:

- `entries`—discarded at the end of this function
- the constructed `*sparse.Vector`—this is the one that lives on.

astralblue committed:
Make SortEntries sort in-place

This is also in line with the sort.Sort() convention (in place).

Suggested by: web123pro

astralblue committed:
Empty out inline local trust/pre-trust after copy

This is so that we avoid unused duplicates in memory.

Spotted by: web123pro
